### ISSUE TITLE
3.2. Particles disappearing when using the scrollwheel outside of running window

### ISSUE THREAD

**Issue by golddotasksquestions** [godotengine/godot#34713]
**Godot version:** 3.2 beta4

**OS/device including version:** Win7

**Issue description:**
When moving the mouse outside of the running window and using the scrollwheel, particles disappear. Does not happen if mouse is inside the running window. There is no input map assigned to the mouse scrollwheel. No code using the scrollwheel scancode. The issue is present when running the project from the editor as well as when running the exported project.
![particles_scrollwheel](https://user-images.githubusercontent.com/47016402/71607561-a73cb700-2b7a-11ea-9e9e-a74ee7a79c48.gif)

**Steps to reproduce:**
Run the minimal project and scroll inside vs outside the running window.

**Minimal reproduction project:**
[particle_test_scrollwheel.zip](https://github.com/godotengine/godot/files/4011745/particle_test_scrollwheel.zip)

**Comment by FlashDaggerX** [comment]
I cannot reproduce this issue on **3.2.beta.custom_build.5fa190513** (commit 5fa190513)

**OS:** Arch Linux, kernel 5.4.6-arch3-1
**Graphics:** Intel Integrated Graphics UHD 620 (Whiskey Lake)

**Comment by golddotasksquestions** [comment]
@FlashDaggerX The frame drop is just due to screen recording. When I don't screen record, then there is absolutely no frame dropping with this minimal project regardless of what I do.

**Comment by Xrayez** [comment]
Seems like particles visibility simply turns off when they go outside screen/viewport, that's an issue I'm experiencing for quite some time, for instance when a projectile with a smoke trail goes outside the screen. I think what is needed is calculating the bounding box of all particles before switching visibility completely, but since it's GPU accelerated not sure how feasible the solution might be.

**Comment by golddotasksquestions** [comment]
@Xrayez  If that's the case, why does are the particles then only disappearing when I use the scrollwheel outside of the viewport?

**Comment by Xrayez** [comment]
I couldn't reproduce either in latest build btw, it just may be connected to the issue in some way, not sure how scrollwheel could affect this. 👀

**Comment by Calinou** [comment]
> I think what is needed is calculating the bounding box of all particles before switching visibility completely, but since it's GPU accelerated not sure how feasible the solution might be.

You can enlarge the AABB manually using **Drawing > Visibility Rect** in the inspector.

However, with CPUParticles2D, the AABB is supposed to be computed automatically since Godot 3.1.

**Comment by Xrayez** [comment]
Thanks, yes it did solve my issue, not sure if this was present back in 2.1 last time I checked! I might have also misinterpreted this property as something which clips the drawing instead so I neglected it in the past...

**Comment by golddotasksquestions** [comment]
@Xrayez You got it!
Changing the Visibility Rect changes the margin after which using the scrollwheel hides visibility of the particles.
I still don't understand why this happens, especially why the scrollwheel thing, but as long as you guys know what going on I'm happy.

**Comment by Xrayez** [comment]
According to docs:

![godot-particles-visibility-rect-docs](https://user-images.githubusercontent.com/17108460/71642521-482f8d00-2cb5-11ea-9e98-196931601e16.png)

I wonder why it mentions editor specifically (misleading?). It may happen that editor input may affect particles in some way (with `Sync Scene Changes` option or similar), but I hardly see much connections in the source code.

**Comment by KoBeWi** [comment]
Can anyone still reproduce this bug in Godot 3.2.3 or any later release?

I can't reproduce either btw

**Comment by golddotasksquestions** [comment]
I can still reproduce it with the minimal project above.

**Comment by Calinou** [comment]
I can't confirm this on 3.3.2 on Linux.

**Comment by golddotasksquestions** [comment]
I can still reproduce it with the included MRP on 3.3.2 stabel on Win.

Increasing the Visibility Rect works as a workaround but does not eliminate the issue.

**Comment by Calinou** [comment]
@golddotasksquestions Can you (or anyone else) still reproduce this bug in [Godot 4.0.beta4](https://godotengine.org/article/dev-snapshot-godot-4-0-beta-4) or any later release?

If yes, please ensure that a 4.0 up-to-date Minimal Reproduction Project (MRP) is included in this report (a MRP is a zipped Godot project with the minimal elements necessary to reliably trigger the bug). You can upload ZIP files in an issue comment with a drag and drop.

**Comment by golddotasksquestions** [comment]
@Calinou 

Yes, I can still reproduce this on the same hardware. Also in Godot4 Beta4. I have converted the original MRP to Godot 4 Beta4 and added it to the original post.

While testing this again, I noticed not only the particles are disappearing, but the whole KinematicBody2D/CharacterBody2D the Particles2D are added to. So maybe this is not related to the Particles? I checked the above gif I made years ago, and even there it seems like the Sprite is disappearing, indicating the KB/CB is totally disappearing when scrolling outside, not just the Particles. That something I apparently did not notice before.

So I made two more small test scenes (one with CharacterBody2D and Sprite2D and one with only the Sprite), and the issue was present in both as long as they where moved to the global_mouse_position()

```
extends Sprite2D

func _process(delta):
	global_position = get_global_mouse_position()
```
